aggregator/src/aggregator.rs

```diff
@@ -2751,375 +2650,481 @@ where
+
+    conn.headers_mut()
+        .insert(KnownHeaderName::CacheControl, "max-age=86400");
+
+    // Handle CORS, if the request header is present.
+    if let Some(origin) = conn.request_headers().get(KnownHeaderName::Origin) {
+        // Unconditionally allow CORS requests from all origins.
+        let origin = origin.clone();
+        conn.headers_mut()
+            .insert(KnownHeaderName::AccessControlAllowOrigin, origin);
+    }
 
-    let hpke_config_endpoint = compose_common_wrappers(
-        warp::path("hpke_config"),
-        hpke_config_responding,
-        warp::cors()
-            .allow_any_origin()
-            .allow_method("GET")
-            .max_age(CORS_PREFLIGHT_CACHE_AGE)
-            .build(),
-        response_time_histogram.clone(),
-        "hpke_config",
-    );
+    Ok(EncodedBody::new(
+        hpke_config_list,
+        HpkeConfigList::MEDIA_TYPE,
+    ))
+}
 
-    let upload_responding = warp::put()
-        .and(warp::path!("tasks" / TaskId / "reports"))
-        .and(warp::header::exact(
-            CONTENT_TYPE.as_str(),
-            Report::MEDIA_TYPE,
-        ))
-        .and(with_cloned_value(Arc::clone(&aggregator)))
-        .and(warp::body::bytes())
-        .then(
-            |task_id: TaskId, aggregator: Arc<Aggregator<C>>, body: Bytes| async move {
-                aggregator
-                    .handle_upload(&task_id, &body)
-                    .await
-                    .map(|_| StatusCode::OK)
-            },
+/// Handler for CORS preflight requests to "/hpke_config".
+async fn hpke_config_cors_preflight(mut conn: Conn) -> Conn {
+    conn.headers_mut().insert(KnownHeaderName::Allow, "GET");
+    if let Some(origin) = conn.request_headers().get(KnownHeaderName::Origin) {
+        let origin = origin.clone();
+        let request_headers = conn.headers_mut();
+        request_headers.insert(KnownHeaderName::AccessControlAllowOrigin, origin);
+        request_headers.insert(KnownHeaderName::AccessControlAllowMethods, "GET");
+        request_headers.insert(
+            KnownHeaderName::AccessControlMaxAge,
+            format!("{CORS_PREFLIGHT_CACHE_AGE}"),
         );
+    }
+    conn.set_status(Status::Ok);
+    conn
+}
 
-    let upload_endpoint = compose_common_wrappers(
-        path_matching_filter("reports")?,
-        upload_responding,
-        warp::cors()
-            .allow_any_origin()
-            .allow_method("PUT")
-            .allow_header("content-type")
-            .max_age(CORS_PREFLIGHT_CACHE_AGE)
-            .build(),
-        response_time_histogram.clone(),
-        "upload",
-    );
+/// API handler for the "/tasks/.../reports" PUT endpoint.
+async fn upload<C: Clock>(
+    conn: &mut Conn,
+    (State(aggregator), State(captures), body): (
+        State<Arc<Aggregator<C>>>,
+        State<Captures<'static, 'static>>,
+        Vec<u8>,
+    ),
+) -> Result<Status, Arc<Error>> {
+    validate_content_type(conn, Report::MEDIA_TYPE).map_err(Arc::new)?;
```

Comment:
We could write a handler-wrapper function, similar to `instrumented`, that could provide this.

Reply:
I think a plain handler would work in this case, it could set the status to 400 and set `halted` if the header is missing or incorrect. Then, we could change the `instrumented(api(...))` handlers inside the router to `(check_header("..."), instrumented(api(...)))`. Whether to do this check in the `trillium` world as a handler, or in the `trillium-api` world with traditional error handling is largely a question of style, I think.